### DISCUSSION TITLE
bash shell language kernel support

### DISCUSSION THREAD

**Discussion by myusrn** [microsoft/vscode-jupyter#9395] [New Ideas]
**Is your feature request related to a problem? Please describe.**
When doing offensive security penetration testing work there is a need to document thinking intermixed with bash shell commands and code used to enumerate, exploit and escalate privileges in a given environment.  This allows others to make sense of what you did and easily step through the process on their system to reproduce the results.  

I see this as being very similar to how when doing data science work you need to document thinking intermixed with R or Python interpreter code that captures how you arrived at your results and allow others to easily reproduce them.

**Describe the solution or feature you'd like**
I would like to have a vscode jupyter notebook story for a bash shell language kernel option to enable authoring notebooks where code cells execute bash commands in a localhost or localhost wsl [ windows subsystem for linux ] distro or remote over ssh distro bash shell terminal.

**Describe alternatives you've considered**
I've done searches to see if anyone has defined a manual way of hacking the vscode jupyter notebook experience to support a bash shell language kernel and haven't found any solution.

**Additional context**
None at this time.  Asked [this](https://github.com/microsoft/azuredatastudio/issues/18752) similar feature question over in the azure data studio repo given it appears to be vscode derivative natively focused on jupyter notebook authoring and execution scenarios.

**Comment by IanMatthewHuff** [comment]
@myusrn VS Code's notebook support should be able to work with existing jupyter kernels. If you install something like [bash kernel](https://pypi.org/project/bash_kernel) that should show up as a kernel option in the notebook interface. Have you tried this already and it's not working?

**Reply by myusrn** [reply]
Hi @IanMatthewHuff thanks for the response and pointer to this option.  I have not tried that [bash kernel](https://pypi.org/project/bash_kernel/) option.  I am guessing that i need to enable it in both my windows miniconda3 and windows wsl kali-linux / ubuntu distro install python environments to cover vscode notebook bash language kernel support for vscode instance launched from windows as well as from windows wsl distro terminal.  Trying now . . .

**Reply by DonJayamanne** [reply]
Optionally you could also use `%%shell bash` magic, which happens to be built into the Python kernel (but now you need to install Python on all machines along with Bash).
See here https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-script

**Reply by myusrn** [reply]
@DonJayamanne thanks for the alternative option.  I tried the `%%bash` [cell magics script](https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-script) option, alias for `%%script bash`, and that worked. It didn't require any additional steps in both vscode launched from windows host with a C:\Users\\\<username>\Miniconda3 python install and vscode launched from a wsl kali-linux and ubuntu distro environment with `sudo apt install python3` install. Not sure yet if this comes with any significant caveats.

@IanMatthewHuff i followed the steps for enabling [bash kernel](https://pypi.org/project/bash_kernel/) support and in the `To use it, run one of:` steps the vscode jupyter notebook editing experience didn't expose a bash language kernel selection after the bash_kernel package and module had been installed.  The `jupyter console --kernel bash` command worked didn't work in terminal of vscode launched from windows host but did work in terminal of vscode launched from wsl kali-linux and ubuntu distro environment. It didn't appear to enable bash language kernel behavior in notebook cells but rather a new bash shell session on top of one that had already been active in the terminal and with no way to exit using `exit` or ctrl+c commands.

**Reply by IanMatthewHuff** [reply]
@myusrn So I just tried the first steps for installing the bash kernel:

```
pip install bash_kernel
python -m bash_kernel.install
```

After running those two steps (+ restarting vscode as we cache kernels that we find on startup) I do now have a bash kernel available. By clicking the kernel selection in the top right of the notebook file (where it says .debugEnv for my example below) in the drop down I have the Bash kernel that was installed.

<img width="1070" alt="image" src="https://user-images.githubusercontent.com/812783/158876039-b971c1d1-40ca-492b-938e-d036123b5294.png">

After picking that I seem to have a reasonable bash notebook interface (note the language for each cell was automatically changed to Shell Script, and the kernel in the top right is changed to Bash). Which I believe is what you were looking for. Does this help at all?

<img width="1154" alt="image" src="https://user-images.githubusercontent.com/812783/158876322-b4c816d7-bbc6-4c60-add9-3b279714ff3a.png">

**Reply by myusrn** [reply]
@IanMatthewHuff thanks for the followup clarification and additional details.  Yes in both my vscode launched from windows as well as vscode launched from wsl distro terminal i was able to see the bash script Jupyter kernel option after restarting vscode.  Selecting it instead of python allowed me to successfully run bash script commands in notebook code cells. Screen grabs provided below.  I see from your screen grabs that this same mechanism should allow me to enable other language kernel options as well such as c# and powershell which is useful.  

q1. Are you aware of whether or not there was some article published that talks about this vscode kernel extensibility support or is it making use of some standard issue jupyter notebook kernel extensibility story that is well documented?

q2. Unrelated to the main issue here so i can raise in new issue thread, but the next step in vetting this option for using vscode notebook for documenting and capturing repro steps of offensive security exploits. Does vscode jupyter notebook authoring support have way to automate uploading to url retrievable location and inserting reference to screen grabbed image into notebook markdown similar to how the github issues authoring pane does when you ctrl+v paste from clipboard into markdown vs having to drop screen grab into file on local disk and enable local reference to it in markdown cell?

![image](https://user-images.githubusercontent.com/3043446/158880893-d4cfe65e-7e9a-481c-b676-592c07257a3e.png)

![image](https://user-images.githubusercontent.com/3043446/158880953-bce49b2d-7812-4619-b5ea-0e5bcc7af3b5.png)

![image](https://user-images.githubusercontent.com/3043446/158880856-ccf8eedc-6f00-45c1-a800-b833b3a06cea.png)
